### PR TITLE
next: decrease surface area between httpcli and http.Client...

### PR DESCRIPTION
**Pull request by jdef**
...and abstract the HTTP round-trip function

### CODE REVIEW

**Comment by tsenart** [comment]
Why not use the RoundTripper interface? That was the purpose of it in the
first place.

File: httpcli/http.go [hunk 1]
```diff
@@ -79,8 +98,11 @@ func (c *Client) Do(m encoding.Marshaler) (encoding.Decoder, io.Closer, error) {
 	req.Header.Set("Content-Type", c.codec.MediaTypes[0])
 	req.Header.Set("Accept", c.codec.MediaTypes[1])
 
-	res, err := c.cli.Do(req)
+	res, err := c.do(req)
 	if err != nil {
+		if res != nil && res.Body != nil {
```
Comment (tsenart) [inline httpcli/http.go]: I don't get this. When there is an error, res should be always nil. This invariant must be part of the API.

**Comment by jdef** [comment]
The semantics of the RoundTripper API are different than that of
http.Client.Do, specifically w/ respect to error handling. I wanted to be
very clear w/ respect to such expectations and felt that using the
RoundTripper API would be confusing in this regard.

File: httpcli/http.go (hunk 1 above)
Comment (jdef) [inline httpcli/http.go]: There is nothing in the API docs for http.Client.Do that says that if err
!= nil then res will be nil.

https://golang.org/pkg/net/http/#Client.Do

Or, am I mis-reading?

**Comment by jdef** [comment]
related w/ respect to headers and "Connection: Close" https://issues.apache.org/jira/browse/MESOS-3601

**Pull request merged**